[DOCS] Edits to audit filtering overview (#25676)
hashicorp/vault: website/content/docs/concepts/filtering/audit.mdx

the predicate expression and not write to the new device. However, the test
message **will** still be written to your fallback audit device.

You can determine whether the test message should appear in the sink for the
fallback or the newly enabled audit device based on the following property
table, which are common to all default test messages.

Property      | Value
------------- | ----------------
`mount_point` | empty
`mount_type`  | empty
`namespace`   | empty
`operation`   | `update`
`path`        | `sys/audit/test`

## `filter` properties for audit devices

Filters can only reference the following properties of an audit entry:

Property      | Example                             | Description
-------------  ------------------------------------ | --------------------------
`mount_point` | `mount_point == auth/oidc`          | Log all entries for the `auth/oidc` mount point
`mount_type`  | `mount_type == kv`                  | Log all entries from `kv` plugins
`namespace`   | `namespace != \"\"`                 | Log all entries **not** in the root namespace
`operation`   | `operation == read`                 | Log all read operations

> `path`        | `path == auth/approle/login`        | Log all activity against the AppRole login path

Author: We should use whichever notation will make readers most successful when they inevitably copy/paste the examples. Does that means adding `\"` to all our examples? (I'm not super familiar with the nuances of JSON pointer notation)

  devices without configuring an audit fallback may lead to gaps in your audit
  logs.
  
  **Always** test your audit configuration in a non-production environment
  before deploying filters to production. And make sure to read the
  [Vault security model](/vault/docs/internals/security) and
  [filtering overview](/vault/docs/concepts/filtering) to familiarize yourself
  with Vault auditing and filtering basics before enabling filtered audit 
  devices.

</Warning>

Once you enable an audit device with a filter, every audit entry Vault sends to
that audit device is compared to the predicate expression in the filter. Only
audit entries that match the filter are written to the audit log for the device.

When you enable filtered audit devices, the behavior of any existing audit
device and the behavior of new audit devices that **do not** use filters remain
unchanged. 

## Fallback auditing devices

Filtering adds flexibility to your auditing workflows, but filtering also adds
complexity that can lead to entries missing from your logs by mistake. For
example, writing audit entries to one device for `(N < 10)` and another device

> for `(N > 10)`would exclude audit entires where `(N == 10)`, which may not be

Reviewer: ```suggestion
for `(N > 10)`would exclude audit entries where `(N == 10)`, which may not be
```

the intended behavior.

> The fallback audit device saves all audit entires that would otherwise get

Reviewer: ```suggestion
The fallback audit device saves all audit entries that would otherwise get
```

1. You can configure filtering when enabling one of the following supported audit device types:
    - [file](/vault/docs/audit/file)
    - [socket](/vault/docs/audit/socket)
    - [syslog](/vault/docs/audit/syslog)
1. You can only designate one auditing fallback device.

## Filtering and test messages

By default, Vault sends a test message to the audit device when you enable it.
Depending on how you configure your filters, the default test message may fail
the predicate expression and not write to the new device.

You can determine whether the test message should appear in the sink for the 
newly enabled audit device based on the following property
table, which are common to all default test messages.

Property      | Value
------------- | ----------------
`mount_point` | empty
`mount_type`  | empty
`namespace`   | empty
`operation`   | `update`
`path`        | `sys/audit/test`

## `filter` properties for audit devices

Filters can only reference the following properties of an audit entry:

Property      | Example                             | Description
------------- | ----------------------------------- | --------------------------

> `mount_point` | `mount_point == auth/oidc`          | Log all entries for the `auth/oidc` mount point
> `mount_type`  | `mount_type == \"kv-v2\"`           | Log all entries from `kv-v2` plugins
> `namespace`   | `namespace != admin/`               | Log all entries **not** in the admin namespace
> `operation`   | `operation == read`                 | Log all read operations
> `path`        | `path == auth/approle/login`        | Log all activity against the AppRole login path

Reviewer: We should encourage users to escape quote the values (I think you've mentioned this in a comment already to make life easier when folks copy/paste).

```suggestion
`mount_point` | `mount_point == \"auth/oidc\"`          | Log all entries for the `auth/oidc` mount point
`mount_type`  | `mount_type == \"kv-v2\"`           | Log all entries from `kv-v2` plugins
`namespace`   | `namespace != \"admin/\"`               | Log all entries **not** in the admin namespace
`operation`   | `operation == \"read\"`                 | Log all read operations
`path`        | `path == \"auth/approle/login\"`        | Log all activity against the AppRole login path
```

<Tip title="Root namespaces are unnamed">

>   Most namespace strings **must** end with a trailing slash to match correctly.
>   But the root namespace is technically unnamed and only matches to an empty

Reviewer: ```suggestion
  Non-root namespace paths **must** end with a trailing slash (`/`) to match correctly.
  But the root namespace does not have a path and only matches to an empty
```

  string. To match to the root namespace in your filter use `\"\"`. For example,

>   `namespace != \"\"` matches any audit entry **not** in the root namespace.

Reviewer: ```suggestion
  `namespace != \"\"` matches any audited request **not** in the root namespace.
```

</Tip>

## A practical example

Assume you already have an audit file called `vault-audit.log` but you want to
filter your audit entries and persist all the key/value (`kv`) type events to a
specific audit log file called `kv-audit.log`.

To filter the events:

1. Enable a `file` audit device with a `mount_type` filter:

>   ```shell-session
>   vault audit enable                \
>     -path kv-only                   \
>     file                            \
>     file_path=/logs/kv-audit.log    \
>     filter="mount_type == kv"
>   ```

Reviewer: I swapped the last two options around so they both end with the `file_path` (and escaped the `kv`).

```suggestion
  ```shell-session
  vault audit enable                \
    -path kv-only                   \
    file                            \
    filter="mount_type == \"kv\""   \
    file_path=/logs/kv-audit.log
  ```
```

1. Enable a fallback device:

>   ```shell-session
>   vault audit enable                \
>     fallback=true                   \
>     -path=my-fallback               \
>     -description="fallback device"  \
>     file file_path=/tmp/kv-audit.fallback.log 
>   ```

Reviewer: `fallback` has to come at the end where the kv config options go.

```suggestion
  ```shell-session
  vault audit enable                \
    -path=my-fallback               \
    -description="fallback device"  \
    file                            \
    fallback=true                   \
    file_path=/tmp/kv-audit.fallback.log
  ```
```

1. Confirm the audit devices are enabled:
  ```shell-session
    vault audit list --detailed
  ```
1. Enable a new `kv` secrets engine called `my-kv`:
  ```shell-session
    vault secrets enable -path my-kv kv-v2
  ```
1. Write secret data to the `kv` engine:
  ```shell-session

>     vault kv put my-kv my_secret=always_angry

Reviewer: For `kv-v2` the syntax is slightly different (sorry).

```suggestion
    vault kv put -mount=my-kv my_secret the_value=always_angry
```

  ```

> The `/var/kv-audit.log` now includes two entries: one for the command that
> enabled `my-kv` and one for the command that wrote a secret to `my-kv`. Entries
> for the other commands are captured by the fallback device. And the original
> audit file, `vault-audit.log` continues to capture all audit events.

Reviewer: There are 4 entries as there's a request and a response for each action (sorry).

```suggestion
The `/var/kv-audit.log` now includes four entries in total: a request and response
entry for the command that enabled `my-kv` and another request and response entry
for the command that wrote a secret to `my-kv`. Entries for the other commands
are captured by the fallback device. And the original audit file, `vault-audit.log`
continues to capture all audit events.
```
